GH-41179: [Docs] Documentation for Dissociated IPC Protocol (#41180)
apache/arrow: docs/source/format/DissociatedIPC.rst

    it is fully standardized.

Rationale
=========

The :ref:`Arrow IPC format <format-ipc>` describes a protocol for transferring
Arrow data as a stream of record batches. This protocol expects a continuous
stream of bytes divided into discrete messages (using a length prefix and
continuation indicator). Each discrete message consists of two portions:

* A `Flatbuffers`_ header message
* A series of bytes consisting of the flattened and packed body buffers (some
message types, like Schema messages, do not have this section)
  - This is referred to as the *message body* in the IPC format spec.

For most cases, the existing IPC format as it currently exists is extremely efficient:

* Receiving data in the IPC format allows zero-copy utilization of the body
buffer bytes, no deserialization is required to form Arrow Arrays
* An IPC (Feather) file can be memory-mapped because it is location agnostic
and the bytes of the file are exactly what is expected in memory.

However, there are use cases that aren't handled by this:

* Constructing the IPC record batch message requires allocating a contiguous
chunk of bytes and copying all of the data buffers into it, packed together

> back-to-back. You currently cannot zero-copy **create** IPC messages.

Reviewer: Well, you _sort of_ can, if you just pass your networking/IO layer a list of buffers (what we do with gRPC in Flight), with padding done by inserting pointers to preallocated padding bufferes. (Of course, internally, somewhere in the networking layer is probably copying, but that's not necessarily under your control anyways.)

Author: funny enough, that's pretty much how the scatter-gather I/O works for UCX and such. I'll change the wording to be "It's extremely difficult to zero-copy **create** IPC messages". That work? :)

* If the Arrow data is located in a shared-memory location, there is no standard
way to share the handle to the shared-memory across processes or transports that
allow for remote memory accessing.
* Arrow data located on a non-CPU device (such as a GPU) cannot be sent using
Arrow IPC without having to copy the data back to the host device or copying
the flatbuffer metadata bytes into device memory.
  - By the same token, receiving IPC messages into device memory would require

>   performing a copy of the flatbuffer metadata back to the host CPU device.

Reviewer: Might be worth explicitly stating that this is because IPC colocates/interleaves data and metadata?

       can be sent across separate pipes/streams/transports.

       If a sequence number reaches ``UINT32_MAX``, it should be allowed to roll over as
       it is unlikely there would be enough unprocessed messages waiting to be processed
       that would cause an overlap of sequence numbers.

       The sequence number serves two purposes: To identify corresponding metadata and 
       tagged body data messages and to ensure we do not rely on messages having to arrive
       in order. A client should use the sequence number to correctly order messages as
       they arrive for processing.

   Backpressure
       *Currently* this proposal does not specify any way to manage the backpressure of
       messages to throttle for memory and bandwidth reasons. For now, this will be 
       **transport-defined** rather than lock into something sub-optimal. 

       As usage among different transports and libraries grows, common patterns will emerge
       that will allow for a generic, but efficient, way to handle backpressure across
       different use cases.

       .. note::
          The current usage and examples have been tested using UCX and libfabric transports

>           so far, but that's all.

Reviewer: is this meant to say something more like, it is intended to work with other transports too?

Author: The first bullet point in the `Goals` section states that it is intended to be a generic protocol not tied to any particular transport. I figured I didn't need to repeat that here, but i'll add it to the note.

Protocol Description
====================

A reference example implementation utilizing `libcudf`_ and `ucx`_ can be found at 
https://github.com/zeroshade/cudf-flight-ucx.

Requirements
------------

A transport implementing this protocol **MUST** provide two pieces of functionality:

* Message sending
  - Delimited messages (like gRPC) as opposed to non-delimited streams (like plain TCP 
  without further framing).
  - Alternatively, a framing mechanism like the `encapsulated message format <ipc-message-format>`
  for the IPC protocol can be used while leaving out the body bytes.
* Tagged message sending
  - Sending a message that has an attached little-endian, unsigned 64-bit integral tag
  for control flow. A tag like this allows control flow to operate on a message whose body
  is on a non-CPU device without requiring the message itself to get copied off of the device.

URI Specification
-----------------

When providing a URI to a consumer to contact for use with this protocol (such as via 

> the `Location URI for Flight <flight-location-uris>`), the URI should specify a scheme

Reviewer: ```suggestion
the `Location URI for Flight <flight-location-uris>`_), the URI should specify a scheme
```
think the syntax is slightly different

  if desired by the server.

After sending the last tagged IPC body message, the server should maintain the connection and wait
for tagged ``<free_data>`` messages. The structure of these ``<free_data>`` messages is simple:
1 or more unsigned, little-endian 64-bit integers which indicate the addresses/offsets that can
be freed. 

Once there are no more outstanding addresses to be freed, the work for this connection is complete.

Client Sequence
---------------

A client for this protocol needs to asynchronously handle both the data and metadata streams of
messages which may either both come from the same server or different servers. Below is a flowchart
showing how a client might handle the metadata and data streams:

.. figure:: ./DissociatedIPC/ClientFlowchart.mmd.svg

#. First the client sends a tagged message using the ``<want_data>`` value it was provided in the
URI as the tag, and the opaque ID as the body.
  * If the metadata and data servers are separate, then a ``<want_data>`` message needs to be sent
  separately to each. 
  * In either scenario, the metadata and data streams can be processed concurrently and/or asynchronously
  depending on the nature of the transports.

> #. For each **untagged** message the client receives in the metadata stream:

Reviewer: How does this actually work in the stated motivation of being able to use Flight RPC for metadata only? gRPC only supports delimited messages, not tagged messages. You'd need some way to initiate a gRPC request alongside a UCX connection, or something.

Author: That would follow the pattern of separate transports / streams for the data and metadata, which would have had two URIs (one for each). The gRPC delimited, untagged messages would be the metadata messages as described, while UCX would send the tagged data messages.

.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. _dissociated-ipc:

========================
Dissociated IPC Protocol
========================

.. warning::

    Experimental: The Dissociated IPC Protocol is experimental in its current
    form. Based on feedback and usage the protocol definition may change until
    it is fully standardized.

Rationale
=========

The :ref:`Arrow IPC format <format-ipc>` describes a protocol for transferring
Arrow data as a stream of record batches. This protocol expects a continuous
stream of bytes divided into discrete messages (using a length prefix and
continuation indicator). Each discrete message consists of two portions:

* A `Flatbuffers`_ header message
* A series of bytes consisting of the flattened and packed body buffers (some
  message types, like Schema messages, do not have this section)
  - This is referred to as the *message body* in the IPC format spec.

For most cases, the existing IPC format as it currently exists is extremely efficient:

> * Receiving data in the IPC format allows zero-copy utilization of the body

Reviewer 2: You're right. I withdraw this.

  buffer bytes, no deserialization is required to form Arrow Arrays

> * An IPC (Feather) file can be memory-mapped because it is location agnostic

Reviewer 2: Ah, "IPC file format file"?

  and the bytes of the file are exactly what is expected in memory.

However, there are use cases that aren't handled by this:

> * Constructing the IPC record batch message requires allocating a contiguous
>   chunk of bytes and copying all of the data buffers into it, packed together
>   back-to-back. It's exceedingly difficult to zero-copy **create** IPC messages.

Reviewer 3: ```suggestion
* Constructing the IPC record batch message requires allocating a contiguous
  chunk of bytes and copying all of the data buffers into it, packed together
  back-to-back. This pessimizes the common case of wrapping existing, directly
  consumable data into an IPC message.
```

* If the Arrow data is located in a shared-memory location, there is no standard
  way to share the handle to the shared-memory across processes or transports that
  allow for remote memory accessing.

> * Arrow data located on a non-CPU device (such as a GPU) cannot be sent using

Reviewer 3: IIUC sending in this context refers to sharing memory directly between GPUs. If so then this could emphasize that this is a special case of the previous point which applies even *within* a process, where colocation of header and data should be easiest to relax.

  Arrow IPC without having to copy the data back to the host device or copying
  the flatbuffer metadata bytes into device memory.
  - By the same token, receiving IPC messages into device memory would require
    performing a copy of the flatbuffer metadata back to the host CPU device. This
    is due to the fact that the IPC stream interleaves data and metadata across a
    single stream.

This protocol is intended to attempt to solve these use cases in an efficient manner.

Goals
-----

* Define a generic protocol for passing Arrow IPC data, not tied to any particular
  transport, that also allows for utilizing non-CPU device memory, shared memory, and
  newer "high performance" transports such as `ucx`_ or `libfabric`_.
* Allow for using :ref:`Flight RPC <flight-rpc>` purely for control flow by separating
  the stream of IPC metadata from IPC body bytes
  - This allows for the data in the body to be kept on non-CPU devices (like GPUs)
    without expensive Device -> Host copies.

Definitions
-----------

> .. glossary::

Author: hmm, okay. I was using it for the formatting mostly, but that's fine. I can change this to use something else instead of "glossary".

   IPC Metadata
       The flatbuffer message bytes that encompass the header of an Arrow IPC message
  
   Tag

>        A ``uint64`` value used as an ID for a message. Specific bits can be masked to

Reviewer 4: Is it little-endian?
Also, is it really an "id" (which is generally expected to be unique)?

Author: it is used for identification and control flow. it's not guaranteed to be unique in the general sense as some protocols use them to identify message *types* rather than identify individual messages. Can you think of a better word to use here instead? since we don't want to give the impression that it is intended to be "unique"

Reviewer 4: Something like "a value that determines how the body of a message is to be interpreted".

Perhaps also add some implementation context, such as:
"""
Some protocols, such as UCX, have built-in support for such tag values and will provide them in CPU memory even though the body of the message may reside on a non-CPU device.
"""

       that would cause an overlap of sequence numbers.

       The sequence number serves two purposes: To identify corresponding metadata and 
       tagged body data messages and to ensure we do not rely on messages having to arrive
       in order. A client should use the sequence number to correctly order messages as
       they arrive for processing.

   Backpressure
       *Currently* this proposal does not specify any way to manage the backpressure of
       messages to throttle for memory and bandwidth reasons. For now, this will be 
       **transport-defined** rather than lock into something sub-optimal. 

       As usage among different transports and libraries grows, common patterns will emerge
       that will allow for a generic, but efficient, way to handle backpressure across
       different use cases.

       .. note::
          While the protocol itself is transport agnostic, the current usage and examples 
          only have been tested using UCX and libfabric transports so far, but that's all.

Protocol Description
====================

A reference example implementation utilizing `libcudf`_ and `ucx`_ can be found at 

> https://github.com/zeroshade/cudf-flight-ucx.

Author: I've filed https://github.com/apache/arrow-experiments/pull/28 to add it to `arrow-experiments` and i'll update the link here when that gets merged.

Requirements
------------

A transport implementing this protocol **MUST** provide two pieces of functionality:

* Message sending
  - Delimited messages (like gRPC) as opposed to non-delimited streams (like plain TCP 
    without further framing).
  - Alternatively, a framing mechanism like the `encapsulated message format <ipc-message-format>`
    for the IPC protocol can be used while leaving out the body bytes.
* Tagged message sending
  - Sending a message that has an attached little-endian, unsigned 64-bit integral tag
    for control flow. A tag like this allows control flow to operate on a message whose body
    is on a non-CPU device without requiring the message itself to get copied off of the device.

URI Specification
-----------------

When providing a URI to a consumer to contact for use with this protocol (such as via 
the `Location URI for Flight <flight-location-uris>`_), the URI should specify a scheme

> like *ucx://* or *fabric://*, that is easily identifiable. In addition, the URI should

Reviewer 4: The `//` is not part of the scheme, it's part of the path (the colon either isn't technically part of the scheme, but I guess it makes the text more easily understood).
```suggestion
like *ucx:* or *fabric:*, that is easily identifiable. In addition, the URI should
```

    transport schemes that get used with it.

* ``want_data`` - **REQUIRED** - uint64 integer value
  - This value should be used to tag an initial message to the server to initiate a
    data transfer. The body of the initiating message should be an opaque binary identifier
    of the data stream being requested (like the ``Ticket`` in the Flight RPC protocol)
* ``free_data`` - **OPTIONAL** - uint64 integer value
  - If the server might send messages using offsets / addresses for remote memory accessing
    or shared memory locations, the URI should include this parameter. This value is used to
    tag messages sent from the client to the data server, containing specific offsets / addresses
    which were provided that are no longer required by the client (i.e. any operations that
    directly reference those memory locations, such as copying the remote data into local memory,
    have been completed).
* ``remote_handle`` - **OPTIONAL** - base64-encoded string
  - When working with shared memory or remote memory, this value indicates any required
    handle or identifier that is necessary for accessing the memory.

>     + Using UCX, this would be an *rkey* value

Reviewer 4: Uh... if this parameter is transport-specific, then should it really be documented here? Also, does it need to be a query parameter? What does an UCX URI look like?

Author: the interpretation of what the parameter means is transport-specific. The documentation here is essentially stating examples of how you would use this parameter for certain transports. 

There is no real standard for a "UCX URI". You only use the host/IP to connect to a server, the address can also be provided via a serialized opaque Worker Address object.

    + With CUDA IPC, this would be the value of the base GPU pointer or memory handle,

>       and subsequent addresses would be offsets from this base pointer.

Reviewer 4: This is... weird. Why don't we just transfer absolute pointers into GPU space?

Reviewer 4: Ah ok, sorry. Is it useful to talk specifically about "offsets"? How does it work when accessing such data: do we explicitly let sum the base pointer and offset, or do we invoke some CUDA API?

Author: That would be up to the consumer. The entire purpose here is to be flexible enough for however the transport / consumer wants to consume the data. The only reason why i mention "offsets" is because it's possible for the values to either be full raw pointers or to be offsets from a base pointer. @kkraus14 could probably better describe the gpu IPC base pointer + offset thing than I can.

Reviewer 5: You explicitly add the offsets to the base pointer. The reason for needing this is that when you try to open CUDA IPC handles, it always works against base pointers of allocations where you can't point to the middle of an allocation as far as I know. Combine that with memory pools being super common for GPU libraries / frameworks / applications, where it's really common to have something like a single base allocation that has millions of suballocations within it and wanting to share one of those suballocations via IPC.

Protocol Description
====================

There are two possibilities that can occur:

> 1. The streams of metadata and body data are across separate connections

Reviewer 4: ```suggestion
1. The streams of metadata and body data are sent across separate connections
```

.. figure:: ./DissociatedIPC/SequenceDiagramSeparate.mmd.svg

2. The streams of metadata and body data are sent simultaneously across the
   same connection

.. figure:: ./DissociatedIPC/SequenceDiagramSame.mmd.svg

Server Sequence
---------------

There can be either a single server handling both the IPC Metadata stream and the
Body data streams, or separate servers for handling the IPC Metadata and the body
data. This allows for streaming of data across either a single transport pipe or
two pipes if desired.

Metadata Stream Sequence
''''''''''''''''''''''''

> The standing state of the server is waiting for a **tagged** message with the specified

Reviewer 4: Specified by whom? This isn't clear.

Author: I'll make this more clear. Ideally the server specifies its `want_data` tag, and then just needs to make sure the value is propagated however the URI is being sent to the client.

``<want_data>`` tag value to initiate a transfer. The body of that message will contain an
opaque, binary identifier to indicate a particular dataset / data stream to send.

.. note::

  For instance, the **ticket** that was passed with a *FlightInfo* message would be
  the body of this message. Because it is opaque, it can be anything the server wants

>   to use. The URI and identifier do not need to be given to the client via Flight RPC,

Reviewer 4: Which identifier is this talking about?

Author: the opaque binary identifier that is sent in the body of the message to identify a particular dataset or data stream

Reviewer 4: Ah, sorry. I should have read more carefully.

  but could come across from any transport or protocol desired.

Upon receiving a ``<want_data>`` request, the server *should* respond by sending a stream
of messages consisting of the following:

* A 5-byte prefix
  - The first byte of the message indicates the type of message, currently there are only
    two allowed message types (more types may get added in the future):
    0) End of Stream

>     1) Flatbuffer IPC Metadata Message

Reviewer 4: Flatbuffers

>   - the next 4-bytes are a little-endian, 32-bit integer indicating the sequence number of

Reviewer 2: ```suggestion
  - the next 4-bytes are a little-endian, unsigned 32-bit integer indicating the sequence number of
```

    the message. The first message in the stream (**MUST** always be a schema message) **MUST**
    have a sequence number of ``0``. Each subsequent message **MUST** increment the number by 
    ``1``.

> * The full Flatbuffer bytes of an Arrow IPC header

Reviewer 4: Flatbuffers

Reviewer 4: Also, what would the Arrow IPC header represent exactly?

Author: the Flatbuffers Arrow IPC header is the metadata necessary for interpreting the buffers. i.e. the structure and nodes of a given message, along with the metadata of length, null_count, offset, etc. for each array etc.

Author: Since we link to the IPC encapsulation documentation earlier, I didn't feel it necessary to re-explain that here

Reviewer 4: Hmm, ok. I was asking because the term "header" can be a bit under-defined.

> After sending the last metadata message, the server **MUST** indicate the end of the stream

Reviewer 4: You could perhaps explain why there may be several metadata messages.

by sending a message consisting of **exactly** 5 bytes:

* The first byte is ``0``, indicating an **End of Stream** message
* The last 4 bytes are the sequence number (4-byte, unsigned integer in little-endian byte order)

Data Stream Sequence
''''''''''''''''''''

If a single server is handling both the data and metadata streams, then the data messages
**should** begin being sent to the client in parallel with the metadata messages. Otherwise,
as with the metadata sequence, the standing state of the server is to wait for a **tagged**
message with the ``<want_data>`` tag value, whose body indicates the dataset / data stream

> to send to the client.

Reviewer 4: To make the two cases more similar, do we want separate `want_metadata` and `want_data` tags?

Author: I don't think so personally. That would introduce confusion for the scenario where they want a single server to be able to provide both the metadata and body data streams (which currently only requires a single message to initiate both streams) where either we would need to make the protocol more "chatty" (requiring an extra message) in that situation or it would be unclear which (if any) of these tags could be left out for that scenario. 

I also prefer keeping it to a single `want` value, rather than requiring two values.

Reviewer 4: Ok, thanks for the clarification.

For each IPC message in the stream of data, a **tagged** message **MUST** be sent on the data
stream if that message has a body (i.e. a Record Batch or Dictionary message). The 
:term:`tag <Tag>` for each message should be structured as follows:

> * The *least significant* 4-bytes (bits 0 - 31) of the tag should be the 32-bit, little-endian sequence 

Reviewer 2: ```suggestion
* The *least significant* 4-bytes (bits 0 - 31) of the tag should be the unsigned 32-bit, little-endian sequence 
```

Reviewer 4: "4 bytes", not "4-bytes"

  number of the message.
* The *most significant* byte (bits 56 - 63) of the tag indicates the message body **type** as an 8-bit
  unsigned integer. Currently only two message types are specified, but more can be added as
  needed to expand the protocol:
  0) The body contains the raw body buffer bytes as a packed buffer (i.e. the standard IPC

>      format body bytes)

Reviewer 2: Can we use both of IPC streaming format and file format?
(I think that we can use only IPC streaming format.)

Reviewer 4: The file format is just the stream format + a footer for direct access. Here the footer is probably pointless, given that we're talking about a stream of data.

  1) The body contains a series of unsigned, little-endian 64-bit integer pairs to represent
     either shared or remote memory, schematically structured as
    - The first two integers (e.g. the first 16 bytes) represent the *total* size (in bytes)
      of all buffers and the number of buffers in this message (and thus the number of following
      pairs of ``uint64``)

>     - Each subsequent pair of ``uint64`` values are an address/offset followed the length of

Reviewer 2: It seems that this document uses both `A/B` style and `A / B` style: 

```suggestion
    - Each subsequent pair of ``uint64`` values are an address / offset followed the length of
```

Reviewer 4: "followed by"?

>       that particular buffer.

Reviewer 2: Need to fix these markup: http://crossbow.voltrondata.com/pr_docs/41180/format/DissociatedIPC.html#data-stream-sequence

* All unspecified bits (bits 32 - 55) of the tag are *reserved* for future use by potential updates
  to this protocol. For now they **MUST** be 0.

.. note::

  Any shared/remote memory addresses that are sent across **MUST** be kept alive by the server
  until a corresponding tagged ``<free_data>`` message is received. If the client disconnects
  before sending any ``<free_data>`` messages, it can be assumed to be safe to clean up the memory
  if desired by the server.

> After sending the last tagged IPC body message, the server should maintain the connection and wait

Reviewer 4: "maintain the connection" implies the server may have to send some keepalive messages? Or does it simply mean "keep the connection open"?

Author: that would depend on the transport being used. If the transport is able to just "keep the connection open" without sending keepalive messages, then that works. If the transport requires sending keepalive messages to keep the connection, then that's what should be done. 

This protocol is intended to be transport agnostic so it's not prescribing *how* to maintain the connection

for tagged ``<free_data>`` messages. The structure of these ``<free_data>`` messages is simple:

> 1 or more unsigned, little-endian 64-bit integers which indicate the addresses/offsets that can

Reviewer 4: "one or more"

be freed. 

> Once there are no more outstanding addresses to be freed, the work for this connection is complete.

Reviewer 4: What does this mean? The client cannot send further `want_data` messages?

Author: I'll change it to say `the work for this stream is complete`. As it should be fine for a client to send further `want_data` messages on the same connection without necessarily requiring a new connection be set up

Client Sequence
---------------

> A client for this protocol needs to asynchronously handle both the data and metadata streams of

Reviewer 4: Do you mean "simultaneously" or "concurrently" rather than "asynchronously"?

Author: good point, i'll fix that

* If the Arrow data is located in a shared memory location, there is no standard
  way to share the handle to the shared-memory across processes or transports that
  allow for remote memory accessing, such as UCX.
* Arrow data located on a non-CPU device (such as a GPU) cannot be sent using
  Arrow IPC without having to copy the data back to the host device or copying
  the Flatbuffers metadata bytes into device memory.
  
  * By the same token, receiving IPC messages into device memory would require
    performing a copy of the Flatbuffers metadata back to the host CPU device. This
    is due to the fact that the IPC stream interleaves data and metadata across a
    single stream.

This protocol attempts to solve these use cases in an efficient manner.

Goals
-----

* Define a generic protocol for passing Arrow IPC data, not tied to any particular
  transport, that also allows for utilizing non-CPU device memory, shared memory, and
  newer "high performance" transports such as `UCX`_ or `libfabric`_.
* Allow for using :ref:`Flight RPC <flight-rpc>` purely for control flow by separating
  the stream of IPC metadata from IPC body bytes
  

>   * This allows for the data in the body to be kept on non-CPU devices (like GPUs)

Reviewer 3: Nit: formatting makes this looks like a subpoint of Flight RPC, but I think it should be a subpoint of the other

* The *least significant* 4-bytes (bits 0 - 31) of the tag should be the unsigned 32-bit, little-endian sequence 
  number of the message.
* The *most significant* byte (bits 56 - 63) of the tag indicates the message body **type** as an 8-bit
  unsigned integer. Currently only two message types are specified, but more can be added as
  needed to expand the protocol:
  
  0) The body contains the raw body buffer bytes as a packed buffer (i.e. the standard IPC
     format body bytes)
  1) The body contains a series of unsigned, little-endian 64-bit integer pairs to represent
     either shared or remote memory, schematically structured as
  
     * The first two integers (e.g. the first 16 bytes) represent the *total* size (in bytes)
       of all buffers and the number of buffers in this message (and thus the number of following
       pairs of ``uint64``)
  
     * Each subsequent pair of ``uint64`` values are an address / offset followed the length of
       that particular buffer.

* All unspecified bits (bits 32 - 55) of the tag are *reserved* for future use by potential updates
  to this protocol. For now they **MUST** be 0.

> 

Reviewer 3: It might be worthwhile to add some graphical clarification of tag structure https://github.com/apache/arrow/blob/1982c2523f5791a5b742697185fd04ee580b464d/docs/source/format/Columnar.rst?plain=1#L268
